Add leading silence detection and removal logic

gexgd0419 opened the pull request:
<!-- Please read and fill in the following template, for an explanation of the sections see:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md
Please also note that the NVDA project has a Citizen and Contributor Code of Conduct which can be found at https://github.com/nvaccess/nvda/blob/master/CODE_OF_CONDUCT.md. NV Access expects that all contributors and other community members read and abide by the rules set out in this document while participating or contributing to this project. This includes creating or commenting on issues and pull requests.

Please initially open PRs as a draft.
When you would like a review, mark the PR as "ready for review".
See https://github.com/nvaccess/nvda/blob/master/.github/CONTRIBUTING.md.
-->

### Link to issue number:
Closes #17614

### Summary of the issue:
Some voices output a leading silence part before the actual speech voice. By removing the silence part, the delay between keypress and user hearing the audio will be shorter, therefore make the voices more responsive.

### Description of user facing changes

Users may find the voices more responsive. All voices using NVDA's `WavePlayer` will be affected, including eSpeak-NG, OneCore, SAPI5, and some third-party voice add-ons.

This should only affect the leading silence parts. Silence between sentences or at punctuation marks are not changed, but this may depend on how the voice uses `WavePlayer`.

### Description of development approach

I wrote a header-only library `silenceDetect.h` in `nvdaHelper/local`. It supports most wave formats (8/16/24/32-bit integer and 32/64-bit float wave), and uses a simple algorithm: check each sample to see if it's outside threshold range (currently hard-coded to +/- 1/2^10 or 0.0009765625). It uses template-related code and requires C++ 20 standard.

The `WasapiPlayer` in `wasapi.cpp` is updated to handle silence. A new member function, `startTrimmingLeadingSilence`, and the exported version `wasPlay_startTrimmingLeadingSilence`, is added, to set or clear the `isTrimmingLeadingSilence` flag. If `isTrimmingLeadingSilence` is true, the next chunk fed in will have its leading silence removed. When non-silence is detected, `isTrimmingLeadingSilence` will be reset to false. So every time a new utterance is about to be spoken, `startTrimmingLeadingSilence` should be called.

In `nvwave.py`, `startTrimmingLeadingSilence()` will be called when:
- the player is initialized;
- the player is stopped;
- `idle` is called;
- `_idleCheck` determines that the player is idle.

Usually voices will call `idle` when an utterance is completed, so that audio ducking can work correctly, so here `idle` is used to mark the starting point of the next utterance. If a voice doesn't use `idle` this way, then this logic might be messed up.

As long as the synthesizer uses `idle` as intended, the synthesizer's code doesn't need to be modified to benefit from this feature.

As leading silence can also be introduced by a `BreakCommand` at the beginning of the speech sequence, `WavePlayer` will check the speech sequence first; if there's a `BreakCommand` at the beginning, the leading silence will not be trimmed for the current utterance. To check the exact speech sequence that is about to be spoken, a new extension point, `pre_synthSpeak`, is added in `synthDriverHandler`, which will be invoked just before `SpeechManager` calls `getSynth().speak()`. The existing `pre_speech` is called before `SpeechManager` processes and queues the speech sequence, so `pre_synthSpeak` is needed to provide a more accurate sequence.

When the purpose of a `WavePlayer` is not `SPEECH`, it does not trim the leading silence by default, because of the way `playWaveFile` works (it calls `idle` after every chunk). Users of `WavePlayer` will still be able to enable/disable automatic trimming by calling `enableTrimmingLeadingSilence`, or to initiate trimming manually for the next audio section by calling `startTrimmingLeadingSilence`.

Other possible ways/things that may worth considering (but hasn't been implemented):
- Put silence detection/removal logic in a separate module instead of in `WavePlayer`. The drawback is that every voice synthesizer module needs to be modified to utilize a separate module.
- Use another audio library, such as [PyDub](https://github.com/jiaaro/pydub), to detect/remove silence.
- Add a setting item to turn this feature on or off. (Update: this has been added in advanced settings)
- Add a public API function to allow a synthesizer to opt out of this feature.

### Testing strategy:

These are the delay results I got using 1X speed. "Improved delay" means the delay after applying this PR.

| Voice          | Original Delay | Improved Delay |
| -------------- | -------------- | -------------- |
| eSpeak NG      |      89ms      |      69ms      |
| Huihui OneCore |     192ms      |      74ms      |
| Huihui SAPI5   |     166ms      |      57ms      |

If the speech rate is higher than 1X, the original delay may be shorter, because the leading silence is also shortened during the speed-up process. When there's no leading silence, changing the rate does not affect the delay much.

Considering the margin of error, we can say that the delay of eSpeak NG, OneCore and SAPI5 voices are now at the same level.

### Known issues with pull request:
None

### Code Review Checklist:

<!--
This checklist is a reminder of things commonly forgotten in a new PR.
Authors, please do a self-review of this pull-request.
Check items to confirm you have thought about the relevance of the item.
Where items are missing (eg unit / system tests), please explain in the PR.
To check an item `- [ ]` becomes `- [x]`, note spacing.
You can also check the checkboxes after the PR is created.
A detailed explanation of this checklist is available here:
https://github.com/nvaccess/nvda/blob/master/projectDocs/dev/githubPullRequestTemplateExplanationAndExamples.md#code-review-checklist
-->

- [x] Documentation:
  - Change log entry
  - User Documentation
  - Developer / Technical Documentation
  - Context sensitive help for GUI changes
- [x] Testing:
  - Unit tests
  - System (end to end) tests
  - Manual testing
- [x] UX of all users considered:
  - Speech
  - Braille
  - Low Vision
  - Different web browsers
  - Localization in other languages / culture than English
- [x] API is compatible with existing add-ons.
- [ ] Security precautions taken.

<!-- Please keep the following -->
@coderabbitai summary

gexgd0419 committed:
Add leading silence detection and removal logic

gexgd0419 committed:
Call startTrimmingLeadingSilence in stop and _idleCheck

zstanecic commented:
@gexgd0419 
I am slightly confused... is your change requiring modifications to synth drivers?

gexgd0419 commented:
> is your change requiring modifications to synth drivers?

In the current implementation, no. As long as the synth driver calls `idle` after each utterance, it will work, because the logic is integrated into `WavePlayer`.

But if separating the logic out of `WavePlayer` is preferred, then the synth driver will have to be aware of the separate module.

cary-rowen commented:
I like this very much, I just ran a simple test without any metrics though.
To put it simply, using the Natural voice adapter and sapi5, in the past, NVDA would not have any voice feedback when I pressed keys quickly. Now this situation has been significantly improved.

gexgd0419 committed:
Add changelog entry

zstanecic commented:
@gexgd0419 
@cary-rowen 
I am not so patient... Yes, i wanted to test it from the source, and i did, so there are my findings:
Compared to the current alpha and stable version of NVDA, i see a drastic improvement when it comes to using various speech synths. 
So... to conclude things, it is a very nice work!
It works with any tts, even with RHVoice in the NVDA addon which i am using now.
There was a time, when typing echo slowed me down a bot. Now, i can have turned on typing echo and listening to speech while typing.

rmcpantoja commented:
Hi, I would like to see this from a critical point, having previous experience in TTS stuff:
Some minor delay in these voices don't come in the voices itself; it does after preprocessing steps. For example, text pre-processing, or a text to phoneme backend, language analyzers or other tools, that are applied before synthesize speech with a provided text. Fortunately, the performance of these resources is very few milliseconds, so that wouldn't be a problem.
And, most of the voices are built by already pre-processed audio, if we are talking of unit selection or Hidden Markov Model voices, in the free or comercial field. That means, technically, silences are trimmed before building these voices from their databases. However, a particular case where certain voices are with small silences are the traditional Microsoft Hidden Markov Model voices, which they are no longer supporting it. In this case, your changes are useful for these certain voices that can improve performance during navigation.
But, anyways, let me try it in these days, and retract my comment.

Cheers.

zstanecic commented:
Hi @rmcpantoja 
This statement in your comment is not so true. RHVoice is also a hidden markov model based. 
We don't do anything with silences when building the voices. We need these to detect pauses in the utterances in the training process. 
Also, we use the Google's vad library to detect silence sometimes. 
Ivona tts is a concatenative tts, and Microsoft natural tts is a concatenative one. 
When using such voices, i hear noticeable differences in terms of speed and responsiveness. 
To conclude things, all voices get the benefits from this feature, and we don't have impact on resulting speech, that means tiere are no drawbacks.

mzanm commented:
Hi. This is amazing work, thank you so much.
One issue is that some synthesizers implement the break command by adding silence I believe or similar, for example when enabling 'delayed descriptions of characters on cursor movement' on the speech panel and when using Espeak and possibly other synths. When moving between characters, the phonetic description will be spoken almost immediately, Which is unexpected as it's supposed to be 1000 milliseconds. It's usually way less than that anyway as it gets affected by fast speech rates in some synthesizers, but here it's near-instant no matter the speech rate on affected synths. Interestingly this issue does not occur on OneCore or SAPI5 with the default David voice, but does additionally occur with the Tiflotecnia voices synthesizer.

gexgd0419 commented:
@mzanm I think I found the cause here: `BreakCommand` at the beginning of an utterance. The "delayed descriptions of characters on cursor movement" feature splits the character and its description into two utterances, and puts a `BreakCommand` at the beginning of the second utterance. As a result, the silence introduced by `BreakCommand` is trimmed, because my code cannot distinguish whether the leading silence is from a `BreakCommand`; it doesn't have access to the original command sequence. So in fact, all synthesizers are affected by this issue. OneCore and SAPI5 voices may just have longer "trailing silence" at the end of the first utterance so it seems normal...

zstanecic commented:
@gexgd0419 Some sapi5 synthesizers have longer trailing silences, for example the slovenian Ebralec or finnish Mikropuhe. 
This is normal, and even this feature works like in the Zdsr screen reader.

gexgd0419 commented:
As I haven't found a way to check the command list in WavePlayer without some kind of modification to the synth driver code, I'm going to try a different method: if the leading silence is longer than, say 200ms, then let the silence pass instead of trimming it. Because longer silence is more likely to be introduced by a `BreakCommand`. But even then, silence shorter than 200ms would still be trimmed. What would be an appropriate default value for the time threshold?

zstanecic commented:
@gexgd0419 I am happy to test it. I am now running it from source.

gexgd0419 commented:
@zstanecic I haven't pushed the changes yet...

zstanecic commented:
@gexgd0419 I know. I am just notifying you. By the way... the good thing will be to have configurable silence treshold in advanced settings for experienced users...

gexgd0419 commented:
Yes, maybe a configurable threshold may be better.

But if my code can have access to the command list to check whether there's a `BreakCommand` at the beginning, I wouldn't have to check the silence length.

This is mainly because `WavePlayer` is designed to play audio only, not to process audio, so all the things given to `WavePlayer` is audio-related.

If the logic is put in `WavePlayer`:

- Can affect many synthesizers without requiring changes to their code.
- The code has no access to other information, such as the original speak command list, so the process might not be accurate.
- Might confuse future `WavePlayer` users why outputting silence does not work.

If the logic is put in another module:

- Can make the synthesizers provide whatever information we need, so the silence detection/removal could be more accurate.
- Requires changes to all existing synthesizers. Without changes, synthesizers will still work as usual, but the leading silence won't be trimmed.

I want to know what NV Access thinks about the idea of putting audio processing logic into `WavePlayer`.

mzanm commented:
I have an idea, but not sure if it'll work. Is it possible to have some kind of flag that determines if next utterance should be trimmed then in speech.speech.speak,  could iterate through the speechSequence and if there's a text command or ratehr string that has no BreakCommand before it then it should set that flag and therefore trim leading silence of next speech. This shouldn't require changes to synthesizers possibly. Or maybe you could register to the pre_speech extension point in nvwave and look at the sequence that way... but probably both methods wouldn't work if speech sequences are queued.

gexgd0419 committed:
Add generateSilence in SilenceDetect

gexgd0419 commented:
@mzanm Thank you for the info. But as you said, pre_speech might be too early, as the speech might be queued. Maybe I can modify the SpeechManager and insert the logic before `getSynth().speak(seq)`.

gexgd0419 committed:
Add pre_synthSpeak extension point in synthDriverHandler

This will be called by SpeechManager just before calling getSynth().speak().

The existing extension point `pre_speech` is called before SpeechManager processes the speech sequence, so a new one is added to get the exact speech sequence that will be spoken immediately by the synth driver.

gexgd0419 committed:
Check speech sequence in nvwave.py to see if leading silence is inserted by BreakCommand

gexgd0419 commented:
To check the exact speech sequence that is about to be spoken, I chose to add a new extension point, `pre_synthSpeak`, in `synthDriverHandler`, which will be invoked just before `SpeechManager` calls `getSynth().speak()`.

@mzanm Now the removed silence when using "delayed descriptions of characters on cursor movement" should be fixed.

However, this makes `WavePlayer` depend on speech-related modules.

@SaschaCowley @seanbudd What do you think of the idea of adding audio processing features to `WavePlayer`? It might not be what it's designed to do, but modifying `WavePlayer` is one of the easiest ways to "fix" things for all synthesizers, especially third-party synthesizers.

zstanecic commented:
@gexgd0419,
@mzanm 
@SaschaCowley 
I can confirm that problem with break command is fixed, for synthesizers which use it. 
I think that it is a good approach, because of the fact that drivers do not need to be modified.

gexgd0419 committed:
Add pre_synthSpeak in developerGuide

jcsteh commented:
I haven't looked at the code here, but for what they're worth, here are some brief thoughts in response to the description and comments:

1. I think it's reasonable to have WavePlayer handle audio processing, and this is how I would have implemented it too. There is already some prior art here: audio ducking gets handled by WavePlayer. This was done precisely to ensure that drivers just get it "for free".
2. That said, I don't love the coupling between WavePlayer and speech. However, I do understand the problem you're trying to solve, and while I wish I had a better solution, I haven't come up with one yet. :(
3. At the very least, I think any coupling with speech should only be done when the "purpose" of the audio is AudioPurpose.SPEECH. Otherwise, we're making decisions about non-speech audio based on speech sequences sent to a synth, which seems wrong.

gexgd0419 commented:
> At the very least, I think any coupling with speech should only be done when the "purpose" of the audio is AudioPurpose.SPEECH. Otherwise, we're making decisions about non-speech audio based on speech sequences sent to a synth, which seems wrong.

@jcsteh That makes sense. But, should I prevent silence trimming altogether for non-speech audio? Or should non-speech audio also be trimmed, just prevent the decision based on the speech sequence?

gexgd0419 committed:
Enable trimming by default for speech only

gexgd0419 committed:
Fix

seanbudd commented:
@AKJUS - are you a real human?

zstanecic commented:
Interesting... how it's going with this?

SaschaCowley commented on nvdaHelper/local/silenceDetect.h:
```diff
@@ -0,0 +1,323 @@
+		if constexpr (std::is_floating_point_v<SampleType>)
+			return SampleType(1);
+		else
+			return (std::numeric_limits<SampleType>::max)() >> ((sizeof(SampleType) - bytesPerSample) * 8);
```
Can we add an explanatory comment for clarity?

SaschaCowley commented on nvdaHelper/local/silenceDetect.h:
```diff
@@ -0,0 +1,323 @@
+		SampleType smp;
+		memcpy(&smp, p, bytesPerSample);
+		smp = Fmt::signExtend(smp);
+		// this sample is out of range, so the leading silence starts at the previous sample
```
This is the end of the leading silence right? I think this would be clearer
```suggestion
		// this sample is out of range, so the previous sample is the final sample of leading silence.
```

SaschaCowley commented on nvdaHelper/local/silenceDetect.h:
```diff
@@ -0,0 +1,323 @@
+		// this sample is out of range, so the trailing silence starts at the next sample
+		if (smp < minValue || smp > maxValue)
+			return size - (p - waveData) - bytesPerSample;
+	} while (p != waveData);
```
Can we use `>` instead of `!=`?
```suggestion
	} while (p > waveData);
```

SaschaCowley commented on nvdaHelper/local/silenceDetect.h:
```diff
@@ -0,0 +1,323 @@
+	return len;
+}
+
+inline std::vector<unsigned char> generateSilenceBytes(const WAVEFORMATEX* wfx, size_t size) {
+	std::vector<unsigned char> wave;
+	size -= size % wfx->nBlockAlign;
+	callByWaveFormat(wfx, [=, &wave](auto fmtTag) {
+		using Fmt = decltype(fmtTag);
+		constexpr auto bytesPerSample = Fmt::bytesPerSample;
+		constexpr auto zeroPoint = Fmt::zeroPoint();
+		if constexpr (zeroPoint == 0) {
+			wave.assign(size, 0);
+		} else if constexpr (bytesPerSample == 1) {
+			wave.assign(size, zeroPoint);
+		} else {
+			wave.assign(size, 0);
+			unsigned char *p = wave.data();
+			unsigned char *pEnd = p + size;
+			for (; p != pEnd; p += bytesPerSample) {
+				memcpy(p, &zeroPoint, bytesPerSample);
+			}
+		}
+	});
+	return wave;
+}
+
+inline std::vector<unsigned char> generateSilenceMs(const WAVEFORMATEX* wfx, unsigned int milliseconds) {
+	return generateSilenceBytes(wfx, (size_t)wfx->nAvgBytesPerSec * milliseconds / 1000);
+}
+
```
Are these needed for this PR?

gexgd0419 replied:
At some point I thought those were needed, but in the current implementation, those are not needed. It might be useful if we want to detect and control other kinds of silence in the future. Should I remove them?

gexgd0419 replied:
At first I thought it should monitor the silence duration and give up trimming the silence if the silence is long enough, because a longer leading silence is more likely to be inserted intentionally. But at the time it decided the silence is too long to be trimmed, it has already "eaten up" some silence, so it should generate some silence to compensate. As I later turned to check the speech queue directly to see if the leading silence is intentionally inserted, those goes unused.

gexgd0419 replied:
As I haven't thought of a possible usage as for now, I will revert the change.

Strictly speaking, trailing silence detection code is also unused. Should I remove that too?

I'm not sure if "removing trailing silence" is also desired. But in order to implement "removing trailing silence" in a stream, we may have to use the `generateSilence` code.

SaschaCowley reviewed (changes requested):
This mostly looks really good to me, nice work @gexgd0419

SaschaCowley commented:
While I think this should be fine most of the time, I still think it would be appropriate to be able to turn it off. Can we get a setting, either in advanced settings, or in speech settings if we think it's something ordinary users are likely to want to turn off?

gexgd0419 commented on nvdaHelper/local/silenceDetect.h:
```diff
@@ -0,0 +1,323 @@
+		}
+	}
+
+	static constexpr SampleType signExtend(SampleType smp) {
```
`signExtend` is to sign-extend a sample to the full size of `SampleType`, so that its sign will be correct.

The code intends to support 24-bit wave samples. When reading a 24-bit sample, we can only read 3 bytes, and store it in a 4-byte integer. However, after copying the 3 bytes, the highest byte of the 4-byte integer remain untouched. It has to be sign-extended to 4-byte, because the sign bit is the most significant bit in the highest byte, or else further comparisons will not work.

Assume that a 24-bit sample of decimal value -16 is read, which is 0xFFFFF0 in hexadecimal. We copy the 3 bytes to a 4-byte integer variable which was zero, so now it becomes 0x00FFFFF0. Notice the highest zero byte wasn't touched. If it is now interpreted as a 4-byte integer, we will get 16777200 in hexadecimal, because it is treated as a positive number without the sign bit set. So here it first gets shifted 8-bits left to become 0xFFFFF000, which is now treated as negative, then shifted 8-bits right, which will sign-extend if `SampleType` is a signed type, so it becomes 0xFFFFFFF0, the intended value of -16.

When the bit count matches the size of `SampleType`, that is any type other than 24-bit, sign extension does not happen.

SaschaCowley replied:
Oh, of course! 🤦‍♀️ Thanks for very clearly and patiently explaining that, I clearly need to remember my introductory CS classes better haha.

gexgd0419 commented:
As this only affects the leading silence, I don't think that ordinary users will want to turn it off to "get longer delay".

However, depending on the actual implementation of the synth driver, it may be possible that this feature "removes too much silence" when using some third-party synth add-ons. Because the code uses `WavePlayer.idle()` as the signal that an utterance is completed, if a synth driver calls `idle` too many times, the logic might remove more silence than intended.

So I think that adding it in the advanced settings is fine. However I haven't studied how to add a settings item in NVDA yet, so some guidance would be appreciated.

gexgd0419 committed:
Clear (initialize) the smp value before reading samples

gexgd0419 committed:
Add comments for min/max value trimming

gexgd0419 committed:
Add Changes for Developers entry

gexgd0419 committed:
Apply suggestions from code review

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

gexgd0419 committed:
Apply suggestions

gexgd0419 committed:
Revert "Add generateSilence in SilenceDetect"

This reverts commit 8ac53260aff24f967b7909d19c7928a91f8b7f92.

cary-rowen commented:
Hi @gexgd0419 

> I haven't studied how to add a settings item in NVDA yet, so some guidance would be appreciated.

You may only need to add a checkbox in the advanced settings panel, so in summary, there are three steps:

1. **Define the configuration specification and default values in [configSpec.py](https://github.com/nvaccess/nvda/blob/master/source/config/configSpec.py):**  
   It should be a boolean type. By default, the behavior should enable this improvement.
2. **Implement the UI in [settingsDialogs.py](https://github.com/nvaccess/nvda/blob/master/source/gui/settingsDialogs.py):**  
   Add a checkbox in the advanced settings panel for the user to toggle. You can use `wx.CheckBox` to create the checkbox and bind it to the configuration item you just defined.
3. **Use the new configuration in your code:**  
   Reference the new configuration setting like this:
   ```python
   if config.conf["general"]["enableImprovement"]:
       pass
   ```

gexgd0419 committed:
Add config item speech.trimLeadingSilence in advanced settings

gexgd0419 committed:
Allow disabling by user

SaschaCowley commented on source/nvwave.py:
```diff
@@ -481,6 +481,8 @@ def enableTrimmingLeadingSilence(self, enable: bool) -> None:
 
 	def startTrimmingLeadingSilence(self, start: bool = True) -> None:
 		"""Start or stop trimming the leading silence from the next audio chunk."""
+		if not config.conf["speech"]["trimLeadingSilence"]:
+			return  ## disabled by user
```
I think this may have unintended consequences. For instance, if trimming leading silence is enabled, and then the user disables it, they will need to restart NVDA for the change to take effect. Instead, I think we should only enable trimming silence if this is true when initialising
https://github.com/gexgd0419/nvda/blob/c2f98f563d4fd775767f76b0f23e2842c968a48d/source/nvwave.py#L297
and then re-initialise speech if this changes  (see https://github.com/gexgd0419/nvda/blob/c2f98f563d4fd775767f76b0f23e2842c968a48d/source/gui/settingsDialogs.py#L3169-L3175 for an example)

SaschaCowley commented on user_docs/en/userGuide.md:
```diff
@@ -3321,6 +3321,11 @@ This option enables behaviour which attempts to cancel speech for expired focus
 In particular moving quickly through messages in Gmail with Chrome can cause NVDA to speak outdated information.
 This functionality is enabled by default as of NVDA 2021.1.
 
+##### Trim leading silence in speech audio {#AdvancedSettingsTrimLeadingSilence}
```
We try to avoid encoding GUI structure in anchors in case the GUI gets reorganised
```suggestion
##### Trim leading silence in speech audio {#TrimLeadingSilenceSpeech}
```
(You'll also need to update https://github.com/gexgd0419/nvda/blob/c2f98f563d4fd775767f76b0f23e2842c968a48d/source/gui/settingsDialogs.py#L3758)

SaschaCowley commented on user_docs/en/userGuide.md:
```diff
@@ -3321,6 +3321,11 @@ This option enables behaviour which attempts to cancel speech for expired focus
 
+##### Trim leading silence in speech audio {#AdvancedSettingsTrimLeadingSilence}
+
+This option enables NVDA to trim the silence at the beginning of the speech audio, making the voices respond faster.
+This option is enabled by default, and should only affect the silence at the beginning. But if you find that some necessary silence periods are also missing (e.g. pause between two sentences) when using a speech synthesizer add-on, you may try turning off this feature.
```
One sentence per line, plus some clarity improvements.
```suggestion
When enabled, NVDA will remove silence from the start of speech audio, which may improve the responsiveness of some speech synthesizers.
This option is enabled by default, and should only affect the silence at the beginning of speech.
If you find that some necessary silence periods are also missing (e.g. pause between two sentences) when using a speech synthesizer add-on, you may turn this feature off entirely to resolve the issue.
```

SaschaCowley commented on nvdaHelper/local/silenceDetect.h:
```diff
@@ -0,0 +1,300 @@
+	return size;
+}
+
+/**
+ * Return the trailing silence wave data length, in bytes.
+ * Assumes the wave data to be of one channel (mono).
+ * Uses a `WaveFormat` type (`Fmt`) to determine the wave format.
+ */
+template <class Fmt>
+size_t getTrailingSilenceSizeMono(
+	const unsigned char* waveData,
+	size_t size,
+	typename Fmt::SampleType threshold
+) {
+	using SampleType = Fmt::SampleType;
+	constexpr size_t bytesPerSample = Fmt::bytesPerSample;
+
+	if (size < bytesPerSample)
+		return 0;
+
+	constexpr SampleType zeroPoint = Fmt::zeroPoint();
+	const SampleType minValue = zeroPoint - threshold, maxValue = zeroPoint + threshold;
+
+	// Check each sample in reverse order
+	const unsigned char* p = waveData + (size - (size % bytesPerSample));
+	SampleType smp = SampleType();
+	do {
+		p -= bytesPerSample;
+		memcpy(&smp, p, bytesPerSample);
+		smp = Fmt::signExtend(smp);
+		// this sample is out of range, so the trailing silence starts at the next sample
+		if (smp < minValue || smp > maxValue)
+			return size - (p - waveData) - bytesPerSample;
+	} while (p > waveData);
+
+	// The whole data block is silence
+	return size;
+}
+
```
Please remove as unused
```suggestion
```

SaschaCowley commented on nvdaHelper/local/silenceDetect.h:
```diff
@@ -0,0 +1,300 @@
+	return len - len % wfx->nBlockAlign;  // round down to block (channel) boundaries
+}
+
+/**
+ * Return the trailing silence wave data length, in bytes.
+ * Uses a `WAVEFORMATEX` to determine the wave format.
+ */
+inline size_t getTrailingSilenceSize(
+	const WAVEFORMATEX* wfx,
+	const unsigned char* waveData,
+	size_t size
+) {
+	size_t len;
+	if (!callByWaveFormat(wfx, [=, &len](auto fmtTag) {
+			using Fmt = decltype(fmtTag);
+			len = getTrailingSilenceSizeMono<Fmt>(
+				waveData, size, Fmt::defaultThreshold());
+		}))
+		return 0;
+
+	size_t align = wfx->nBlockAlign;
+	len += align - 1;
+	len -= len % align;  // round up to block (channel) boundaries
+	return len;
+}
```
Please remove as unused
```suggestion
```

SaschaCowley reviewed (changes requested):
A few minor changes

SaschaCowley commented:
@gexgd0419 I think the string you went with is great. Thanks for providing guidance, @cary-rowen

gexgd0419 committed:
Remove unused code

gexgd0419 committed:
Remove GUI structure from name "AdvancedSettingsTrimLeadingSilence"

gexgd0419 committed:
Update user_docs/en/userGuide.md

Co-authored-by: Sascha Cowley <16543535+SaschaCowley@users.noreply.github.com>

gexgd0419 committed:
Reload synth when settings changes

gexgd0419 committed:
Add copyright header for silenceDetect.h

pre-commit-ci[bot] committed:
Pre-commit auto-fix

SaschaCowley commented on user_docs/en/userGuide.md:
```diff
@@ -3321,10 +3321,11 @@ This option enables behaviour which attempts to cancel speech for expired focus
 This functionality is enabled by default as of NVDA 2021.1.
 
-##### Trim leading silence in speech audio {#AdvancedSettingsTrimLeadingSilence}
+##### Trim leading silence in speech audio {#TrimLeadingSilence}
```
Just in case we elect to trim silence in other situations for some reason
```suggestion
##### Trim leading silence in speech audio {#TrimLeadingSilenceSpeech}
```

SaschaCowley commented on source/gui/settingsDialogs.py:
```diff
@@ -3755,7 +3755,7 @@ def __init__(self, parent):
 		label = _("Trim leading silence in speech audio")
 		self.trimLeadingSilenceCheckBox = speechGroup.addItem(wx.CheckBox(speechBox, label=label))
-		self.bindHelpEvent("AdvancedSettingsTrimLeadingSilence", self.trimLeadingSilenceCheckBox)
+		self.bindHelpEvent("TrimLeadingSilence", self.trimLeadingSilenceCheckBox)
```
```suggestion
		self.bindHelpEvent("TrimLeadingSilenceSpeech", self.trimLeadingSilenceCheckBox)
```

SaschaCowley committed:
Change anchor

SaschaCowley reviewed (approved):
Thanks @gexgd0419, nice work!

Qchristensen reviewed (approved):
Great work, thank you.

wmhn1872265132 commented:
Since this feature is now allowed to be disabled, can it be merged during the 2025.1 development cycle?

zstanecic commented:
I also agree with @wmhn1872265132 
This feature is a great improvement, and if we have more time for testing in alpha, apparently we have, it will be good to have it there. 
I am running this from source and i see a massive improvement with various synths.

seanbudd commented on source/nvwave.py:
```diff
@@ -206,6 +209,7 @@ def isInError() -> bool:
 
 
 wasPlay_callback = CFUNCTYPE(None, c_void_p, c_uint)
+_isLeadingSilenceInserted: bool = False
```
could this be made into an instance variable of `WasapiWavePlayer`. I'm concerned with this being global and multiple streams being possible.

gexgd0419 replied:
As for now, there is only one SpeechManager, which always uses the current synthesizer. So there is only one active speech sequence, which is examined by any player that is playing at the moment.

Even if it is possible that multiple players can be used at the same time, they are checking the same (current) speech sequence, and the `_isLeadingSilenceInserted` will not be different between players.

gexgd0419 replied:
Done.

seanbudd commented on source/nvwave.py:
```diff
@@ -530,6 +561,18 @@ def _idleCheck(cls):
 fileWavePlayerThread: threading.Thread | None = None
 
 
+def _onPreSpeak(speechSequence: SpeechSequence):
```
can this be made function of `WasapiWavePlayer` for similar reasoning as multiple instances. Alternatively, pass in `_isLeadingSilenceInserted` from a wrapper function in `WasapiWavePlayer`

seanbudd committed:
Update user_docs/en/changes.md

gexgd0419 committed:
Move speech queue checking into WavePlayer class

pre-commit-ci[bot] committed:
Pre-commit auto-fix

gexgd0419 committed:
Remove unused function `WaveFormat::convertFrom`

gexgd0419 commented:
I removed `WaveFormat::convertFrom` because it is not used in the code. Currently, the threshold is hard-coded to be 1 / 2^10 or 0.0009765625, which can be calculated at compile time, so no need to convert custom threshold value at the moment.

Still, I think that I might be over-engineering this a little bit...

LeonarddeR commented:
It looks like this feature is breaking the letter "p" when at the beginning of a speech sequence.

gexgd0419 commented:
@LeonarddeR What is the voice and the speech rate you are using?

LeonarddeR commented:
I'm using espeak wit rate boost on, rate is 30 but I've also seen it on lower rates. I'm using the default max voice with the Dutch language.

beqabeqa473 commented:
When using some synths, like russian newfon, first letter is everytime cut.

zstanecic commented:
@beqabeqa473 @LeonarddeR @gexgd0419 
The problem is not happening with RHVoice.

gexgd0419 commented:
The threshold is hard-coded to 1 / 2^10 or 0.0009765625. Maybe it's not small enough, so some extra audio is trimmed.

Should I make the threshold value adjustable via settings? Or just try another hard-coded value?

zstanecic commented:
Hi!
@gexgd0419 
Make it adjustable. It is best option, so that users can adjust it.

gexgd0419 commented:
@LeonarddeR @beqabeqa473 I think that this may also be related to the volume level in NVDA (not the system volume level). Because the speech audio will first be limited by the NVDA volume level, then pass through my silence detector, then limited by the system volume level. If the volume is low enough before passing through the silence detector then more audio might be trimmed.

Could you test if it would become better when using a higher volume level in NVDA?

gexgd0419 commented:
@Danstiv Confirmed that the first beep wave will be chopped off a little bit, and this does not happen without this feature.

gexgd0419 commented:
Here is my PR #17699 to fix those problems. Please test if it works. @LeonarddeR @beqabeqa473 @Danstiv 

@zstanecic Seems that this issue has nothing to do with the threshold setting, so I made no change to that.
